prow/plugins/releasenote/releasenote.go

```diff
@@ -29,98 +31,272 @@ import (
-	return handle(pc.GitHubClient, pc.Logger, ic)
+func handlePullRequest(pc plugins.PluginClient, pr github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.Logger, &pr)
 }
 
-func handle(gc githubClient, log *logrus.Entry, ic github.IssueCommentEvent) error {
-	// Only consider PRs and new comments.
-	if !ic.Issue.IsPullRequest() || ic.Action != github.IssueCommentActionCreated {
+func handlePR(gc githubClient, log *logrus.Entry, pr *github.PullRequestEvent) error {
+	// Only consider events that edit the PR body.
+	if pr.Action != github.PullRequestActionOpened && pr.Action != github.PullRequestActionEdited {
 		return nil
 	}
+	org := pr.Repo.Owner.Login
+	repo := pr.Repo.Name
 
-	org := ic.Repo.Owner.Login
-	repo := ic.Repo.Name
-	number := ic.Issue.Number
-
-	// Which label does the comment want us to add?
-	var nl string
-	switch {
-	case releaseNoteRe.MatchString(ic.Comment.Body):
-		nl = releaseNote
-	case releaseNoteNoneRe.MatchString(ic.Comment.Body):
-		nl = releaseNoteNone
-	case releaseNoteActionRequiredRe.MatchString(ic.Comment.Body):
-		nl = releaseNoteActionRequired
-	default:
-		return nil
+	prLabels, err := gc.GetIssueLabels(org, repo, pr.Number)
+	if err != nil {
+		return fmt.Errorf("failed to list labels on PR #%d. err: %v", pr.Number, err)
+	}
+
+	labelToAdd := determineReleaseNoteLabel(pr.PullRequest.Body)
+	if labelToAdd == releaseNoteLabelNeeded {
+		if !prMustFollowRelNoteProcess(gc, log, pr, prLabels, true) {
+			ensureNoRelNoteNeededLabel(gc, log, pr, prLabels)
+			return nil
+		}
+		if !hasLabel(releaseNoteLabelNeeded, prLabels) {
+			comment := plugins.FormatResponse(pr.PullRequest.User.Login, releaseNoteBody, releaseNoteSuffix)
+			if err := gc.CreateComment(org, repo, pr.Number, comment); err != nil {
+				log.WithError(err).Errorf("Failed to comment on %s/%s#%d with comment %q.", org, repo, pr.Number, comment)
+			}
+		}
+	} else {
+		//going to apply some other release-note-label
+		ensureNoRelNoteNeededLabel(gc, log, pr, prLabels)
+	}
+
+	// Add the label if needed
+	if !hasLabel(labelToAdd, prLabels) {
+		gc.AddLabel(org, repo, pr.Number, labelToAdd)
+	}
+
+	err = removeOtherLabels(
+		func(l string) error {
+			return gc.RemoveLabel(org, repo, pr.Number, l)
+		},
+		labelToAdd,
+		allRNLabels,
+		prLabels,
+	)
+	if err != nil {
+		log.Error(err)
 	}
 
-	// Only allow authors and org members to add labels.
-	isMember, err := gc.IsMember(ic.Repo.Owner.Login, ic.Comment.User.Login)
+	// Clean up old comments.
+	// If the PR must follow the process and hasn't yet completed the process, don't remove comments.
+	if prMustFollowRelNoteProcess(gc, log, pr, prLabels, false) && !releaseNoteAlreadyAdded(prLabels) {
+		return nil
+	}
+	botName, err := gc.BotName()
 	if err != nil {
 		return err
 	}
+	return gc.DeleteStaleComments(
+		org,
+		repo,
+		pr.Number,
+		func(c github.IssueComment) bool {
+			if c.User.Login != botName {
+				return false
+			}
+			if !strings.Contains(c.Body, releaseNoteBody) &&
```

Comment:
shorten this to:

    return strings.Contains(c.Body, releaseNoteBody) || strings.Contains(c.Body, parentReleaseNoteBody) || c.Body == deprecatedReleaseNoteBody

Reply:
Done.